odoo/addons/base/models/ir_access.py

```diff
@@ -0,0 +1,424 @@
+        """
+        # use an empty context for 'user' to make the domain evaluation
+        # independent from the context
+        return {
+            'user': self.env.user.with_context({}),
+            'time': time,
+            'company_ids': self.env.companies.ids,
+            'company_id': self.env.company.id,
+        }
+
+    @ormcache('model_name', 'operation')
+    def _get_access_groups(self, model_name, operation='read'):
+        """ Return the group expression object that represents the users who
+        can perform ``operation`` on model ``model_name``.
+        """
+        assert operation in OPERATIONS, "Invalid access operation"
+
+        model_id = self.env['ir.model']._get_id(model_name)
+        domain = [
+            ('model_id', '=', model_id),
+            ('group_id', '!=', False),
+            (f'for_{operation}', '=', True),
+            ('active', '=', True),
+        ]
+        accesses = self.sudo().search_fetch(domain, ['group_id'])
+
+        group_definitions = self.env['res.groups']._get_group_definitions()
+        return group_definitions.from_ids(accesses.group_id.ids)
+
+    def _get_allowed_models(self, operation='read') -> set[str]:
+        """ Return all the models for which the given operation is possible. """
+        assert operation in OPERATIONS, "Invalid access operation"
+
+        group_ids = self.env.user._get_group_ids()
+        domain = [('access_ids', 'any', [
+            ('active', '=', True),
+            (f'for_{operation}', '=', True),
+            ('group_id', 'in', group_ids),
+        ])]
+        models = self.sudo().env['ir.model'].search_fetch(domain, ['model'], order='id')
+
+        return set(models.mapped('model'))
+
+    def _make_access_error(self, records, operation: str) -> AccessError:
+        """ Return the exception to be raised in case of access error.
+        Use an empty ``records`` if the current user has no permission at all to
+        perform ``operation`` on the model.
+        """
+        if records:
+            return self._make_record_access_error(records, operation)
+        else:
+            return self._make_model_access_error(records._name, operation)
+
+    def _make_model_access_error(self, model_name: str, operation: str) -> AccessError:
+        operation_error = str(MODEL_ACCESS_ERROR[operation]) % dict(
+            document_kind=self.env['ir.model']._get(model_name).name or model_name,
+            document_model=model_name,
+        )
+
+        groups = self._get_groups_with_access(model_name, operation)
+        if groups:
+            group_info = _(
+                "This operation is allowed for the following groups:\n%(groups_list)s",
+                groups_list="\n".join(f"\t- {group.display_name}" for group in groups),
+            )
+        else:
+            group_info = _("No group currently allows this operation.")
+
+        resolution_info = _("Contact your administrator to request access if necessary.")
```

Comment:
<sub><sup>Should I mention that this is not a funny message? :skull:</sup></sub>